tests/components/tasmota/test_discovery.py

```diff
@@ -0,0 +1,372 @@
+    assert device_entry is not None
+    entity_entry = entity_reg.async_get("switch.beer")
+    assert entity_entry is not None
+
+    state = hass.states.get("switch.beer")
+    assert state is not None
+    assert state.name == "Beer"
+
+    assert (mac, "switch", 0) in hass.data[ALREADY_DISCOVERED]
+
+
+async def test_device_discover(
+    hass, mqtt_mock, caplog, device_reg, entity_reg, setup_tasmota
+):
+    """Test setting up a device."""
+    config = copy.deepcopy(DEFAULT_CONFIG)
+    mac = config["mac"]
+
+    async_fire_mqtt_message(
+        hass,
+        f"{DEFAULT_PREFIX}/{mac}/config",
+        json.dumps(config),
+    )
+    await hass.async_block_till_done()
+
+    # Verify device and registry entries are created
+    device_entry = device_reg.async_get_device(set(), {("mac", mac)})
+    assert device_entry is not None
+    assert device_entry.manufacturer == "Tasmota"
+    assert device_entry.model == config["md"]
+    assert device_entry.name == config["dn"]
+    assert device_entry.sw_version == config["sw"]
+
+
+async def test_device_update(
+    hass, mqtt_mock, caplog, device_reg, entity_reg, setup_tasmota
+):
+    """Test updating a device."""
+    config = copy.deepcopy(DEFAULT_CONFIG)
+    config["md"] = "Model 1"
+    config["dn"] = "Name 1"
+    config["sw"] = "v1.2.3.4"
+    mac = config["mac"]
+
+    async_fire_mqtt_message(
+        hass,
+        f"{DEFAULT_PREFIX}/{mac}/config",
+        json.dumps(config),
+    )
+    await hass.async_block_till_done()
+
+    # Verify device entry is created
+    device_entry = device_reg.async_get_device(set(), {("mac", mac)})
+    assert device_entry is not None
+
+    # Update device parameters
+    config["md"] = "Another model"
+    config["dn"] = "Another name"
+    config["sw"] = "v6.6.6"
+
+    async_fire_mqtt_message(
+        hass,
+        f"{DEFAULT_PREFIX}/{mac}/config",
+        json.dumps(config),
+    )
+    await hass.async_block_till_done()
+
+    # Verify device entry is updated
+    device_entry = device_reg.async_get_device(set(), {("mac", mac)})
+    assert device_entry is not None
+    assert device_entry.model == "Another model"
+    assert device_entry.name == "Another name"
+    assert device_entry.sw_version == "v6.6.6"
+
+
+@pytest.mark.no_fail_on_log_exception
+async def test_discovery_broken(hass, mqtt_mock, caplog, device_reg, setup_tasmota):
+    """Test handling of exception when creating discovered device."""
+    config = copy.deepcopy(DEFAULT_CONFIG)
+    mac = config["mac"]
+    data = json.dumps(config)
+
+    # Trigger an exception when the entity is added
+    with patch(
+        "hatasmota.discovery.get_device_config_helper",
+        return_value=object(),
+    ):
+        async_fire_mqtt_message(hass, f"{DEFAULT_PREFIX}/{mac}/config", data)
+        await hass.async_block_till_done()
+
+    # Verify device entry is not created
+    device_entry = device_reg.async_get_device(set(), {("mac", mac)})
+    assert device_entry is None
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6285789)